addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,764 @@
+        for tax_percentage, tax_percentage_values in tax_details_grouped['tax_details'].items():
+            tax_breakdowns.append({
+                'taxPercentage': tax_percentage['tax_percentage'],
+                'taxableAmount': tax_percentage_values['base_amount_currency'],
+                'taxAmount': tax_percentage_values['tax_amount_currency'],
+                'taxableAmountPos': self.move_type == 'out_invoice',  # For adjustment invoice, the amount should be considered as negative.
+                'taxAmountPos': self.move_type == 'out_invoice',  # Same
+            })
+
+        json_values['taxBreakdowns'] = tax_breakdowns
+
+    def _l10n_vn_edi_lookup_invoice(self):
+        """ Lookup on invoice, returning its current details on SInvoice. """
+        self.ensure_one()
+        invoice_data, error_message = self._l10n_vn_edi_send_request(
+            method='POST',
+            endpoint='InvoiceAPI/InvoiceWS/searchInvoiceByTransactionUuid',
+            params={
+                'supplierTaxCode': self.company_id.vat,
+                'transactionUuid': self.l10n_vn_edi_invoice_transaction_id,
+            },
+        )
+        return invoice_data, error_message
+
+    @api.model
+    def _l10n_vn_edi_send_request(self, method, endpoint, json_data=None, params=None):
+        """ Send a request to the API based on the given parameters. In case of errors, the error message is returned. """
+        url = urllib.parse.urljoin(SINVOICE_API_URL, endpoint)
+        if json_data:
+            headers = {'Content-Type': 'application/json; charset=utf-8'}
+        else:
+            headers = {'Content-Type': 'application/x-www-form-urlencoded;'}
+            params = urllib.parse.urlencode(params)
+
+        if endpoint != '/auth/login':  # in which case we do not yet have the token.
+            access_token, error = self._l10n_vn_edi_get_access_token()
+            headers['Cookie'] = f'access_token={access_token}'
+            if error:
+                return {}, error
+
+        try:
+            response = requests.request(method, url, json=json_data, params=params, headers=headers, timeout=SINVOICE_TIMEOUT)
+            resp_json = response.json()
+            error = None
+            if resp_json.get('code') or resp_json.get('error'):
+                data = resp_json.get('data') or resp_json.get('error')
+                error = _('Error when contacting SInvoice: %s.', data)
+            return resp_json, error
+        except (RequestException, ValueError) as err:
+            return {}, _('Something went wrong, please try again later: %s', err)
+
+    def _l10n_vn_edi_get_access_token(self):
+        """ Return an access token to be used to contact the API. Either take a valid stored one or get a new one. """
+        self.ensure_one()
+        # First, check if we have a token stored and if it is still valid.
+        if self.company_id.l10n_vn_edi_token and self.company_id.l10n_vn_edi_token_expiry > datetime.now():
```

Comment:
Not working when dealing with branches! What if the branch has no EDI configuration but the root company does? I think we should do something similar to the mexican EDI

Reply:
Hm I didn't think of it. In practice your company is setup into sinvoice, so I would expect to generate have one pair of credentials per company.
But anyway, it doesn't do any harm indeed to fallback to the parent company just in case.